docs/dx.rst

```diff
@@ -0,0 +1,246 @@
+     - Not allowed
+     - Not possible
+     - Copied within existing dir
+   * - serverside copy/copytree
+     - Allowed
+     - Not allowed
+     - Allowed
+   * - copytree, target exists as file
+     - Not allowed
+     - Allowed
+     - Allowed
+   * - copytree, target exists as dir
+     - Merges the two directories
+     - Overwrites the dir
+     - Copies inside existing directory.
+       If root folder is moved, project name is used while copying if needed.
+
+Files on DNAnexus
+-----------------
+Files stored on the DX platform are immutable. This is because the files are
+internally stored in AWS while the metadata handling is taken care of by the
+platform. Hence, once a file is uploaded, it cannot be modified.
+
+When one file is copied to another project, only additional metadata is produced,
+while the underlying file on AWS remains the same. This is essential. The same
+file with the same canonical ID will appear in both projects, and can have
+different folder paths. Deleting a file from one project is possible, which
+deletes the metadata and leaves the file untouched in other projects.
+
+DXPath on stor
+--------------
+All `DXPath` instances on stor must have a project. The project is the first part
+of the path and is determined by a trailing ``:``. Thus, paths without a project
+like::
+
+    Path('dx://')
+    OR
+    Path('dx://path/to/file')
+
+can never be initialized on stor. Trying so will result in an error. The one and
+only exception to this scenario is when only a project is mentioned. Thus,::
+
+    Path('dx://myproject:')
+    OR
+    Path('dx://myproject')
+
+both are valid, and essentially refer to the same project path.
+
+Files and directories can have the same name on the DX platform. This is because
+files without extentions and/or directories with extensions are allowed. A
+`DXPath` instance on stor can be ambiguous as a result. For example, the path::
+
+    Path('dx://myproject:/path/to/folder_file')
+
+can refer to a file and/or a folder, since ``folder_file`` can be a folder and/or
+a file. Hence different operations may refer to the different underlying object.
+As an example, `DXPath.listdir` would treat the above path as a folder, while
+`DXPath.stat` would treat the same path as a file.
+
+DNAnexus also allows two objects to have duplicate file names within the same absolute
+path (same project and directory), because the underlying objects are different.
+However, given a virtual path which resolves to two namesake file objects, stor cannot
+distinguish between the two and will raise a `DuplicateError` if operations are made on
+such a virtual path. Such duplicate objects can only be handled through their canonical
+paths on stor. Use the dx-toolkit to determine their unique IDs.
+
+DXPath
+-------
+
+.. automodule:: stor.dx
+    :members:
+
+Behavior of copy and copytree
+-----------------------------
+Since there are directories on DNAnexus, and  a file and a folder with the
```

Comment:
This is handled by the table above right?

Let's call this:

```
Copy and copytree by example
```

as a section name

and use the CLI interface for this (because it's less noisy)

```
Copy and copytree by example
----

copy and copytree behave differently when the target output path exists and is a folder. (this holds DX -> DX and also POSIX -> DX)

| command | output path (no folder) | output path (folder exists) |
| stor cp  myfile.vcf dx://project2:/call | dx://project2:/call| dx://project2:/call/myfile.vcf |
| stor cp -r ./trip-photos dx://newproject:/all | dx://newproject:/all | dx://newproject/all/trip-photos |

Note that if the output path exists and is a file, the file will be *overwritten*
```

Reply:
Changed.